src/System.Net.Sockets/src/System/Net/Sockets/_MultipleConnectAsync.cs

```diff
@@ -277,6 +368,11 @@ private bool Fail(bool sync, Exception e)
         private void SyncFail(Exception e)
         {
             OnFail(false);
+
+            if (!SocketPal.SupportsMultipleConnectAttempts && _lastAttemptSocket != null)
+            {
+                _lastAttemptSocket.Dispose();
+            }
```

Comment:
These three lines seem to be repeated a fair number of times.  Consider refactoring them out into a DisposeLastAttemptSocketIfNecessary, or something like that.